### PR TITLE
Simplify MLLP.Sender/Client api and standardize errors 

### PR DESCRIPTION
**Pull request by starbelly**
- reduce the various send_* functions to send/2,/3,/4 and send_async/2,3
- add auto_reconnect_interval option
- wrap and unwrap payloads by default
- only verify a response with MLLP.Ack if the payload provided is an HL7.Message.t()
- move the initial connection attempt into a handle_continue/2 clause
- normalize (validate) host parameter to a tuple and raise if invalid
- Always log an error when we fail to establish a connection
- Store connection failures in state so we can return the caller the last known error in case
   they try to send a message while the connection is still being established
- Add and standardize on MLLP.Client.Error (exception) for returned errors
- Use Config vs  Mix.Config in config/*
- misc refactors, enhancements, and fixups

### CODE REVIEW

File: lib/mllp/sender.ex
Comment (vikas15bhardwaj) [inline lib/mllp/sender.ex]: I see we currently have some non_hl7 and raw send functions, which I believe are not used normally. I haven't seen anyone requesting non_hl7 messages on MLLP. We do have some use case to receive non_hl7 on MLLP.

@bryanhunter - Do you see any reason we should keep non_hl7 send support?

File: lib/mllp/sender.ex
Comment (vikas15bhardwaj) [inline lib/mllp/sender.ex]: wouldn't we always return ACK message or error? In what cases we return `{:ok, binary()}`

Comment (starbelly) [reply lib/mllp/sender.ex]: Good catch.  That's actually redundant. It should be just `{:ok, String.t()}`  `String.t()` is equivalent to `binary()`

Comment (starbelly) [reply lib/mllp/sender.ex]: > wouldn't we always return ACK message or error? 

I might have misunderstood. We would return a binary (or `String.t())` if verify_reply is false (i.e., send raw and receive raw reply or send %HL7{} but verify_reply is set to false).

File: lib/mllp/sender.ex
Comment (starbelly) [inline lib/mllp/sender.ex]: There's two things I don't like about this function name. 

1. It can be seen as redundant `Sender.send()` maybe `Client.send()` would read better
2. It sends AND receives, which is why I suspect it was `send_and_receive*` before yet, I don't like having an `and` in the function name either. OTOH, we can say, well sometimes sending implies a response and in the case of MLLP/HL7 that's always true. 

Suggestions and thoughts appreciated. What I do like about it, it's nice and short ;)

Comment (vikas15bhardwaj) [reply lib/mllp/sender.ex]: I am ok with name `send`, but `Sender.send()` looks little weird and I like `Client.send()`. It's anyway is a MLLP client sending messages to a server.

File: lib/mllp/sender.ex
Comment (vikas15bhardwaj) [reply lib/mllp/sender.ex]: I take `raw` message here still an HL7 message but as `binary()` rather than type HL7.Message.t(). I think `send` should accept both `binary` and `HL7.Message.t()`. If message received is an `HL7.Message.t()`, then we should always return verified ACK. If we receive `binary()` raw hl7, we return `binary()` response. I think, we don't need `verify_reply` option

Comment (starbelly) [reply lib/mllp/sender.ex]: Ahh yeah, it does, but I failed to add this to the callback spec. Thanks!

File: lib/mllp/sender.ex
Comment (starbelly) [reply lib/mllp/sender.ex]: Let's rename it to Client then :) 👍

File: lib/mllp/sender.ex
Comment (starbelly) [inline lib/mllp/sender.ex:1]: This should not been a delete since I did `git mv`. I'll have to fix this up.

File: lib/mllp/client.ex
Comment (starbelly) [inline lib/mllp/client.ex:1]: For some reason  git is  being stubborn about marking this as a new  file vs moved and  changed. I can try to fix  it, let me know.

File: config/config.exs
Comment (starbelly) [inline config/config.exs:3]: Went ahead and changed this since Mix.Config is deprecated

File: config/test.exs
Comment (starbelly) [inline config/test.exs:1]: Went ahead and changed this since Mix.Config is deprecated

File: lib/mllp/client.ex
Comment (starbelly) [inline lib/mllp/client.ex]: I think this should be moved to another module because it could be tested as a pure function, but I would opt to do that  in another PR.

File: lib/mllp/client.ex
Comment (starbelly) [inline lib/mllp/client.ex]: This should be `defp`, but will wait to make that  change.  This is merely so we don't have to constantly construct the socket address every time we want to log it. 

The caller can override the defaults at startup or on send/*

File: lib/mllp/client.ex
Comment (starbelly) [inline lib/mllp/client.ex:419]: This is a move to standardize on MLLP.Client.Error for what we return (or raise). `:message` is present if the caller (or  us in a future PR) wish to  raise  and provide a decent error message. 

`:type` was changed to `:context` as that's really what we're saying with `:type` previously (i.e., providing a context). 

`:reason` merely holds the raw error if the caller wishes to match on  it or do something with it.

File: lib/mllp/client.ex
Comment (starbelly) [inline lib/mllp/client.ex]: Added mostly for logging  purposes.

File: lib/mllp/client/error.ex
Comment (starbelly) [inline lib/mllp/client/error.ex:1]: See https://github.com/HCA-Healthcare/elixir-mllp/pull/27/files#r770810192

File: test/client_and_receiver_integration_test.exs
Comment (starbelly) [inline test/client_and_receiver_integration_test.exs:1]: Renamed to align  with renaming Sender to Client.

File: test/client_and_receiver_integration_test.exs
Comment (starbelly) [inline test/client_and_receiver_integration_test.exs:291]: Had to change port numbers all the way down. There's a better  way of handling this, namely using port 0 and providing the randomly chosen port to the test. This IMO should  be done in another PR.

File: test/client_and_receiver_integration_test.exs
Comment (starbelly) [inline test/client_and_receiver_integration_test.exs]: This has to be done as per the changes around the case when  socket is `nil`, our return is more consistent, but the charlist (string) varies from OTP ver  to OTP ver (consistent between OTP 23 and 24, but diff on 22). 

I have thoughts on improving all this, but should probably be  done in another PR. But, mainly it's  usually not  advantageous to match on exact strings like this as it makes for brittle tests, but the way the tests have been DRYed up makes working  in a better way of handling this not a small change.

File: test/client_test.exs
Comment (starbelly) [inline test/client_test.exs:1]: Git  doing the same thing here, says the file was added vs renamed / moved.

File: lib/mllp/client.ex
Comment (starbelly) [inline lib/mllp/client.ex:7]: Some  of the names of these options could perhaps be better, but not a big deal either.

File: lib/mllp/client.ex
Comment (starbelly) [inline lib/mllp/client.ex:335]: We now log on a  failure to connect always, we also store the last connect failure error in state in the case the caller tries to send when we the connection is still trying to be established.

File: lib/mllp/client.ex
Comment (starbelly) [inline lib/mllp/client.ex:384]: We optimize for the case that the caller expect MLLP.Client to  wrap the message and expects a reply, and that  reply is to be verified via MLLP.Ack.

File: test/client_and_receiver_integration_test.exs
Comment (vikas15bhardwaj) [inline test/client_and_receiver_integration_test.exs]: Can we use this code instead. It will match on any context value unless matching on any context value is an issue

```
assert match?(
               {:error, %Error{context: _, message: "connection closed"}},
               MLLP.Client.send(sender_pid, "Simple message")
             )
```

File: test/client_and_receiver_integration_test.exs
Comment (vikas15bhardwaj) [reply test/client_and_receiver_integration_test.exs:291]: correct, I also didn't like using so many ports. I tried using only one port but it fails when parallel tests are running in CI. Locally it worked fine

Comment (starbelly) [reply test/client_and_receiver_integration_test.exs:291]: Interesting, I'll give it a second set of eyes later today, may simply be not possible in CI, but we could also just make a  custom random port function in tests ourselves (i.e., get all listening ports, pick  a  port number not being used).

File: test/client_and_receiver_integration_test.exs
Comment (vikas15bhardwaj) [reply test/client_and_receiver_integration_test.exs]: If you like, we can ignore message matching by making following changes to these tests:
```
  describe "client cert validation" do
    setup ctx do
      verify = ctx[:verify] || :verify_peer
      allowed_clients = ctx[:allowed_clients] || []

      tls_options = [
        cacertfile: "tls/root-ca/ca_certificate.pem",
        certfile: "tls/server/server_certificate.pem",
        keyfile: "tls/server/private_key.pem",
        verify: verify
      ]

      transport_opts = %{tls: tls_options}

      {:ok, %{pid: receiver_pid}} =
        MLLP.Receiver.start(
          port: ctx.port,
          dispatcher: MLLP.EchoDispatcher,
          transport_opts: transport_opts,
          allowed_clients: allowed_clients
        )

      client_cert = ctx[:client_cert] || "tls/client/client_certificate.pem"
      keyfile = ctx[:keyfile] || "tls/client/private_key.pem"

      sender_tls_options = [
        verify: :verify_peer,
        cacertfile: "tls/root-ca/ca_certificate.pem",
        certfile: client_cert,
        keyfile: keyfile
      ]

      tls_alert = ctx[:reason] || []

      expected_error_reasons = [:einval, :no_socket, :closed] ++ tls_alert

      on_exit(fn -> MLLP.Receiver.stop(ctx.port) end)

      [
        receiver_pid: receiver_pid,
        sender_tls_options: sender_tls_options,
        expected_error_reasons: expected_error_reasons
      ]
    end

    @tag port: 8161
    @tag verify: :verify_none
    @tag client_cert: ""
    @tag keyfile: ""
    test "does not verify client cert if verify none option is provided on receiver", ctx do
      make_call_and_assert_success(ctx, ctx.ack)
    end

    @tag port: 8162
    @tag client_cert: ""
    @tag keyfile: ""

    @tag reason: [:handshake_failure, :certificate_required]
    test "no peer cert", ctx do
      make_call_and_assert_failure(ctx, ctx.expected_error_reasons)
    end

    @tag port: 8163
    test "accepts a peer cert", ctx do
      make_call_and_assert_success(ctx, ctx.ack)
    end

    @tag port: 8164
    @tag allowed_clients: ["client-1"]
    test "accepts peer cert for allowed client", ctx do
      make_call_and_assert_success(ctx, ctx.ack)
    end

    @tag port: 8165
    @tag allowed_clients: ["client-x", "client-y"]
    test "reject peer cert for unexpected clients", ctx do
      expected_error_reasons = [:einval, :closed]

      log =
        capture_log(fn ->
          make_call_and_assert_failure(ctx, expected_error_reasons)
        end)

      assert log =~ ":fail_to_verify_client_cert"
    end

    @tag port: 8166
    @tag client_cert: "tls/expired_client/client_certificate.pem"
    @tag keyfile: "tls/expired_client/private_key.pem"

    @tag reason: [:certificate_expired]
    test "reject expired peer cert", ctx do
      make_call_and_assert_failure(ctx, ctx.expected_error_reasons)
    end

    @tag port: 8167
    @tag client_cert: "tls/server/server_certificate.pem"
    @tag keyfile: "tls/server/private_key.pem"

    @tag reason: [:handshake_failure]
    test "reject server cert as peer cert", ctx do
      make_call_and_assert_failure(ctx, ctx.expected_error_reasons)
    end

    @tag port: 8168
    @tag allowed_clients: ["client-x", "client-1"]
    test "accept peer cert from multiple allowed clients", ctx do
      make_call_and_assert_success(ctx, ctx.ack)
    end

    defp make_call_and_assert_success(ctx, expected_result) do
      assert expected_result == start_sender_and_send(ctx)
    end

    defp make_call_and_assert_failure(ctx, expected_error_reasons) do
      reason =
        start_sender_and_send(ctx)
        |> case do
          {:error, %Error{reason: {:tls_alert, {reason, _}}}} -> reason
          {:error, %Error{reason: reason}} -> reason
        end

      assert reason in expected_error_reasons
      # assert expected_error_reasons |> Enum.any?(&match?(^reason, &1))
    end

    defp start_sender_and_send(ctx) do
      {:ok, sender_pid} =
        MLLP.Client.start_link("localhost", ctx.port, tls: ctx.sender_tls_options)

      MLLP.Client.send(
        sender_pid,
        HL7.Examples.wikipedia_sample_hl7() |> HL7.Message.new()
      )
    end
  end
```

File: lib/mllp/client.ex
Comment (vikas15bhardwaj) [inline lib/mllp/client.ex]: should be renamed to client_error

Comment (starbelly) [reply lib/mllp/client.ex]: Nice catch  :)

File: lib/mllp/client.ex
Comment (vikas15bhardwaj) [inline lib/mllp/client.ex]: I think, we can rename variable here to `normalized_address`

File: lib/mllp/client.ex
Comment (vikas15bhardwaj) [inline lib/mllp/client.ex]: Should we raise `MLLP.Client.Error` here?

File: lib/mllp/client.ex
Comment (vikas15bhardwaj) [inline lib/mllp/client.ex]: and here too?

File: lib/mllp/client.ex
Comment (starbelly) [reply lib/mllp/client.ex]: I think a  more appropriate error would be ArgumentError, wdyt?

File: lib/mllp/client.ex
Comment (starbelly) [reply lib/mllp/client.ex]: I removed the intermediate var all  together now, but can put back in place if you like.

File: lib/mllp/client.ex
Comment (vikas15bhardwaj) [inline lib/mllp/client.ex]: Should we put this time in Config, so we can have different delays for tests and reduce sleep time in tests

File: lib/mllp/client.ex
Comment (vikas15bhardwaj) [reply lib/mllp/client.ex]: any reason we can't make it private now

File: lib/mllp/client.ex
Comment (starbelly) [reply lib/mllp/client.ex]: Yeah, makes perfect sense to me. Was going to do that in another PR, but  might as  well do it here :)

File: lib/mllp/client.ex
Comment (vikas15bhardwaj) [inline lib/mllp/client.ex]: Are these formatting for logging purpose. Can we not simply do `inspect` on error?

Comment (starbelly) [reply lib/mllp/client.ex]: Yes,  for logging purpose,  I figured it's friendlier (and in some cases  cheaper) to avoid an inspect/1

File: lib/mllp/client.ex
Comment (vikas15bhardwaj) [inline lib/mllp/client.ex]: I always wonder, why we should create multiple processes for same address and port. Does it introduce process leaks Or it limits us for processing the messages. I think leaving it up to caller is OK. What do you think?

File: lib/mllp/client.ex
Comment (vikas15bhardwaj) [inline lib/mllp/client.ex:17]: Couple of points on this:

- Should we have two methods `send_async` and `send`
     - send_async - as the name suggest does not wait for reply but can send reply using a callback. It also takes away `exact_reply` option
     - send - is synchronous and waiting for reply

On `verify_reply` - It's about checking reply if its valid HL7 ACK or not. Should we do it only if payload is `HL7.Message.t()`, otherwise we don't need to do it. And we can remove this flag also.

On `wrap_message` - Should we do it if message is not wrapped without using the flag?

File: lib/mllp/client.ex
Comment (starbelly) [reply lib/mllp/client.ex]: I'd leave it up to the caller. A lot of MLLP servers in the wild from what I've  read are multiplexing.

File: lib/mllp/client.ex
Comment (starbelly) [reply lib/mllp/client.ex:17]: > Should we have two methods `send_async` and `send`
> 
>     * send_async - as the name suggest does not wait for reply but can send reply using a callback. It also takes away `exact_reply` parameter
> 
>     * send - is synchronous and waiting for reply


I  like  that idea a lot, I very much don't like all the options that can be manipulated. The less options the better, but see below. 

> On verify_reply - It's about checking reply if its valid HL7 ACK or not. Should we do it only if payload is HL7.Message.t(), otherwise we don't need to do it. And we can remove this flag also.

That makes sense  to me. I simply didn't know all the cases we would want to support. In that regard, I think `send_async` is not even really needed per the MLLP protocol itself. Can  you think of an actual case where a caller would want to sent and not receive a reply? I  can not. 

> On wrap_message - Should we do it if message is not wrapped without using the flag?

I thought about that, that makes perfect sense to me,  but it comes back to the same question.  Are  there actually any cases where someone would send data that's not enveloped? I think a lof  of the decisions I made around options was merely to support the existing test cases, but I didn't feel great about them.

Comment (vikas15bhardwaj) [reply lib/mllp/client.ex:17]: `wrap_message` -  Most caller shouldn't worry about wrapping message at their end. It's a MLLP protocol thing and I think it should be handled here.
`send_async` - I haven't seen this case either, but since we have its implementation, we can keep it. It may be useful in cases where someone wants to just blast off messages without worrying about ACKs during some sort of test.

Comment (starbelly) [reply lib/mllp/client.ex:17]: Should  we always  unwrap per getting rid of wrap_message option? Makes sense to me.

File: lib/mllp/client.ex
Comment (starbelly) [reply lib/mllp/client.ex]: Yup, we  can

**Comment by vikas15bhardwaj** [comment]
Some minor comments:
- Can we add `send_async` to `ClientContract`
- We should change `:sender` to client in one telemetry event on line 366 in client module
- In multiple tests, we are using `sender` as variable names for client, should we change those?

**Review by vikas15bhardwaj — APPROVED** [review]

**Pull request merged**
